e2e/nomostest/reset.go

```diff
@@ -0,0 +1,513 @@
+		// dependencies that have owners references. Otherwise the reconciler
+		// and dependenencies will be garbage collected before the finalizer
+		// can delete the managed resources.
+		// TODO: Remove explicit Background policy after the reconciler-manager finalizer is added.
+		nt.T.Logf("[RESET] Deleting RepoSync %s", rsNN)
+		if err := nt.Delete(rs, client.PropagationPolicy(metav1.DeletePropagationBackground)); err != nil {
+			return err
+		}
+	}
+	tg := taskgroup.New()
+	for _, item := range rsList {
+		obj := &item
+		nn := client.ObjectKeyFromObject(obj)
+		nt.T.Logf("[RESET] Waiting for deletion of RepoSync %s ...", nn)
+		tg.Go(func() error {
+			return WatchForNotFound(nt, kinds.RepoSyncV1Beta1(), obj.Name, obj.Namespace)
+		})
+	}
+	if err := tg.Wait(); err != nil {
+		return err
+	}
+
+	// Delete any RoleBindings left behind.
+	// For central control, the parent RSync _should_ handle deleting the RB,
+	// but for delegated control and other edge cases clean them up regardless.
+	nt.T.Log("[RESET] Deleting test RoleBindings")
+	var rbs []client.Object
+	for rsNN := range record.Objects {
+		rbs = append(rbs, RepoSyncRoleBinding(rsNN))
+	}
+	// Skip deleting managed RoleBindings
+	rbs, err := findUnmanaged(nt, rbs...)
+	if err != nil {
+		return err
+	}
+	if err := batchDeleteAndWait(nt, rbs...); err != nil {
+		return err
+	}
+
+	// Delete any ClusterRoleBindings left behind.
+	// CRBs are usually only applied if PSP was enabled, but clean them up regardless.
+	nt.T.Log("[RESET] Deleting test ClusterRoleBindings")
+	var crbs []client.Object
+	for rsNN := range record.Objects {
+		crbs = append(crbs, repoSyncClusterRoleBinding(rsNN))
+	}
+	// Skip deleting managed ClusterRoleBindings
+	crbs, err = findUnmanaged(nt, crbs...)
+	if err != nil {
+		return err
+	}
+	if err := batchDeleteAndWait(nt, crbs...); err != nil {
+		return err
+	}
+
+	return deleteRepoSyncClusterRole(nt)
+}
+
+// ResetNamespaces cleans up one or more Namespaces and all their namespaced objects.
+// Use this for cleaning up Namespaces in tests that use delegated control.
+func ResetNamespaces(nt *NT, nsList []corev1.Namespace) error {
+	if len(nsList) == 0 {
+		return nil
+	}
+
+	nt.T.Logf("[RESET] Deleting Namespaces (%d)", len(nsList))
+
+	record := newResetRecord()
+
+	for _, item := range nsList {
+		obj := &item
+		nn := client.ObjectKeyFromObject(obj)
+		record.Objects[nn] = struct{}{}
+		record.ObjectNamespaces[nn.Namespace] = struct{}{}
+
+		// If managed, skip direct deletion
+		if manager, found := obj.GetAnnotations()[string(metadata.ResourceManagerKey)]; found {
+			record.ManagedObjects[nn] = struct{}{}
+			nt.T.Logf("[RESET] Namespace %s managed by %q", nn, manager)
+			continue
+		}
+
+		nt.T.Logf("[RESET] Deleting Namespace %s", nn)
+		if err := nt.Delete(obj, client.PropagationPolicy(metav1.DeletePropagationForeground)); err != nil {
+			return err
+		}
+	}
+	for _, item := range nsList {
```

Comment:
hmm.. If one of the namespaces was managed, won't it not get deleted in the above loop? I assume that means we would fail in this loop

Reply:
> If one of the namespaces was managed, won't it not get deleted in the above loop?
Correct

> I assume that means we would fail in this loop.
Maybe. It depends on how ResetRepoSyncs is used.

For example, when called after ResetRootSyncs, you can assume all the RootSyncs are deleted, and RepoSyncs can't manage namespaces, so your assertion would be true. But if you call ResetRepoSyncs without ResetRootSyncs, then it's possible that other RootSyncs are being torn down asynchronously... 

To really be clean about it, we could wait here for the unmanaged ones, but then we'd have to duplicate the code to wait for all the rest at the end of `Reset`.

Is it worth duplicating the code to keep this method "pure"?

Reply:
Ah right, on second thought I don't think there should ever be any managed namespaces at this point since the `RootSync`/`RepoSync` cleanup happens before this function is invoked. I suppose maybe that means we don't need the special handling of managed namespaces? I assume all of the `RootSync` managed resources should be cleaned up at this point since we wait at the end of `ResetRootSyncs`. Is that right?

Reply:
>  I suppose maybe that means we don't need the special handling of managed namespaces?

I exposed these functions as public so I could call ResetRepoSyncs from the composition test, to replace the cleanup it was doing there, which is before all the other RootSyncs are deleted. It's used there because half the RSyncs are managed and half are unmanaged, but some of the unmanaged RepoSyncs depend on namespaces and rolebindings managed by the RootSync...

So there's a little bit of extra complexity in here to allow that usage that isn't strictly necessary for just the full `Reset` use case.